compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLBufferInstructionsOptimization.cpp

```diff
@@ -0,0 +1,131 @@
+// %read = vector.transfer_read %memref, %mask : memref<..., amdgpu.raw_fat_buffer>
+// becomes
+// %padding = arith.constant dense<0> : vector<1x ... x1x8xbf16>
+// %read = vector.transfer_read %memref : memref<..., amdgpu.raw_fat_buffer> // no mask!
+// %masked_read
+//   = arith.select %0 && ... && %n ? %read : %padding : index,vector<1x ... x1x8xbf16>
+// clang-format on
+class SimplifyMaskVectorTransferRead
+    : public OpRewritePattern<vector::CreateMaskOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::CreateMaskOp maskOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = maskOp.getLoc();
+    auto readOp = dyn_cast<vector::TransferReadOp>(
+        *(maskOp.getResult().getUsers().begin()));
+    if (!readOp) {
+      return failure();
+    }
+
+    auto sourceType = dyn_cast<MemRefType>(readOp.getSource().getType());
+    if (!sourceType) {
+      return failure();
+    }
+    // This optimization only works on fat raw buffers.
+    if (!hasAMDGPUFatRawBufferAddressSpace(sourceType)) {
+      return failure();
+    }
+
+    SmallVector<bool> inBounds = readOp.getInBoundsValues();
+    if (inBounds.size() != sourceType.getRank()) {
+      return failure();
+    }
+
+    SmallVector<Value> maskIndices = maskOp.getOperands();
+    ArrayRef<int64_t> maskShape = maskOp.getResult().getType().getShape();
+    SmallVector<Value> ValuesToAnd;
+    for (auto [idx, i] : llvm::enumerate(maskIndices)) {
+      auto constantValue = getConstantIndex(i);
+      // We only support the pattern for reads that are in-bounds.
+      if (inBounds[idx] != true) {
+        return failure();
+      }
+      if (constantValue) {
+        // We bail-out if we can statically determine that a mask dim is not
+        // full.
+        if (maskShape[idx] != constantValue) {
+          return failure();
+        }
+      } else {
+        // The non-constant dimensions are only allowed to be unit to ensure
+        // an all ones or an all zeros mask.
+        if (maskShape[idx] != 1) {
+          return failure();
+        }
+        ValuesToAnd.push_back(i);
+      }
+    }
+    // There are no non-constant unit dimensions which means this is
+    // a always one mask, such masks should be folded by vector
+    // canonicalizations so we dont handle it.
+    if (ValuesToAnd.size() == 0) {
+      return failure();
+    }
+    auto selectValue = createi1And(loc, ValuesToAnd, rewriter);
+    auto constantValue = rewriter.create<vector::SplatOp>(
+        loc, readOp.getVectorType(), readOp.getPadding());
+
+    auto newReadOp = rewriter.create<vector::TransferReadOp>(
+        loc, readOp.getVectorType(), readOp.getSource(), readOp.getIndices(),
+        readOp.getPadding(), ArrayRef<bool>{inBounds});
+    rewriter.replaceOpWithNewOp<arith::SelectOp>(readOp, selectValue, newReadOp,
+                                                 constantValue);
```

Comment:
Shall we erase the old transfer_read op explicitly? Looks like you are replacing it with the new read op then rely on compiler to remove it.

Reply:
I believe `replaceOpWithNewOp` does the erase for you along with replacing all uses.

Reply:
https://mlir.llvm.org/doxygen/classmlir_1_1RewriterBase.html#afb1c910a57707f518d2b9c903c2bb5bc

Reply:
Just an FYI, now I am doing create + replaceAllUsesWith, which doesnt do the erase. This was need to incorpate suggestion from Jakub to use walkAndApplyPatterns instead of greedy rewrite. I think this is still ok though as the old read is a dead op and gets removed at the next canonicalization.

Reply:
You should still be `eraseOp()` ing the old one, not letting canonicalization handle it

Reply:
What's the harm? I think it's pretty common to let canonicalizon deal with the dead ops. Anyway I personally don't care but it's causing a problem with the walkAndApply that @kuhar wants.